File: backend/app/schemas/payroll_month.py
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, Field, model_validator


class PayrollMonthBlocker(BaseModel):
    code: str
    message: str
    person_id: int | None = None
    work_date: date | None = None
    work_date_end: date | None = None


class PayrollMonthPersonApprovalSummary(BaseModel):
    approved_count: int
    total_count: int


class PayrollMonthPersonApprovalRead(BaseModel):
    person_id: int
    person_name: str
    status: str
    approval_version: int
    approved_at: datetime | None = None
    approved_by_name: str | None = None
    reopened_at: datetime | None = None
    reopened_by_name: str | None = None
    reopen_reason: str | None = None
    blocker_count: int
    blockers: list[PayrollMonthBlocker] = Field(default_factory=list)
    has_blocking_technical_error: bool
    export_ready: bool
    export_status: str
    export_message: str | None = None
    can_approve: bool
    can_reopen: bool


class PayrollMonthStatusRead(BaseModel):
    year: int
    month: int
    status: str
    snapshot_id: int | None = None
    snapshot_version: int | None = None
    locked_at: datetime | None = None
    locked_by_name: str | None = None
    can_lock: bool
    can_reopen: bool
    artifacts_ready: bool
    blockers: list[PayrollMonthBlocker] = Field(default_factory=list)
    person_approval_summary: PayrollMonthPersonApprovalSummary | None = None
    person_approvals: list[PayrollMonthPersonApprovalRead] = Field(default_factory=list)


class PayrollMonthPersonApprovalRequest(BaseModel):
    confirmed: bool
    acknowledged_blocker_count: int = Field(ge=0)


class PayrollMonthPersonReopenRequest(BaseModel):
    reason: str = Field(min_length=1, max_length=1000)

    @model_validator(mode="after")
    def clean_reason(self):
        self.reason = self.reason.strip()
        if not self.reason:
            raise ValueError("Eine Begründung ist Pflicht.")
        return self


class PayrollMonthLockRequest(BaseModel):
    confirmed: bool


class PayrollMonthReopenRequest(BaseModel):
    reason: str = Field(min_length=1, max_length=1000)

    @model_validator(mode="after")
    def clean_reason(self):
        self.reason = self.reason.strip()
        if not self.reason:
            raise ValueError("Eine Begründung ist Pflicht.")
        return self


class PayrollMonthAuditRead(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    snapshot_id: int | None = None
    action: str
    status_before: str
    status_after: str
    reason: str | None = None
    details_json: dict | None = None
    user_id: int | None = None
    created_at: datetime

File: frontend/src/types/payrollMonth.ts
export type PayrollMonthPeriodStatus = "OPEN" | "LOCKED";
export type PayrollMonthPersonApprovalStatus = "OPEN" | "APPROVED";

export type PayrollMonthLockStatus = {
  year: number;
  month: number;
  status: PayrollMonthPeriodStatus;
  approved_person_ids: number[];
};

export type PayrollMonthBlocker = {
  code: string;
  message: string;
  person_id?: number | null;
  work_date?: string | null;
  work_date_end?: string | null;
};

export type PayrollMonthPeriod = {
  year: number;
  month: number;
  status: PayrollMonthPeriodStatus;
  snapshot_id: number | null;
  snapshot_version: number | null;
  locked_at: string | null;
  locked_by_name: string | null;
  can_lock: boolean;
  can_reopen: boolean;
  artifacts_ready: boolean;
  blockers: PayrollMonthBlocker[];
  person_approval_summary: PayrollMonthPersonApprovalSummary | null;
  person_approvals: PayrollMonthPersonApproval[];
};

export type PayrollMonthPersonApprovalSummary = {
  approved_count: number;
  total_count: number;
};

export type PayrollMonthPersonApproval = {
  person_id: number;
  person_name: string;
  status: PayrollMonthPersonApprovalStatus;
  approval_version: number;
  approved_at: string | null;
  approved_by_name: string | null;
  reopened_at: string | null;
  reopened_by_name: string | null;
  reopen_reason: string | null;
  blocker_count: number;
  blocker_fingerprint: string;
  blockers: PayrollMonthBlocker[];
  has_blocking_technical_error: boolean;
  has_remarks: boolean;
  export_ready: boolean;
  export_status: "READY" | "UNAVAILABLE";
  export_message: string | null;
  can_approve: boolean;
  can_reopen: boolean;
};

export type PayrollWeeklyPlan = {
  id: number;
  valid_from: string;
  valid_to: string | null;
  weekday_minutes: number[];
  weekly_minutes: number;
  contract_weekly_minutes: number | null;
  is_confirmed: boolean;
  confirmed_by_name: string | null;
  confirmed_at: string | null;
  note: string | null;
};

export type PayrollWeeklyPlanUpdate = {
  valid_from: string;
  weekday_minutes: number[];
  note?: string | null;
  confirm: true;
};
